### DISCUSSION TITLE
What's the purpose of useEventInput() hook?

### DISCUSSION THREAD

**Discussion by anderson-chong** [edmundhung/conform#347] [Q&A]
I was quite confused the real usage of useInputEvent hook, and I have some unanswered questions:
1. Why use `useInputEvent` if I could just use `useState()` to update the hidden input?
2. When and why to use each control method? E.g. `control.change`, `control.focus` and `control.blur`. Because I noticed in the official examples, not all control methods are used in all integrated components.
3. What's the differences of the `onFocus` handler in `useInputEvent` and the `onFocus` property in the referenced input? (Vice versa for `onChange` and `onBlur`)

**Comment by edmundhung** [comment]
I am sorry that you find it confusing. To answer your question, let me give you a quick overview how Conform understand a "field" first.

As you might know, Conform utilizes `event delegation`. It expects:

- An `input` event to be dispatched with `event.target` points to the input element when the value is changed, and
- A `blur` event to be dispatched with `event.target` also points to the input element whenever it loses the focus 

This is exactly what browsers does with all native inputs. However, this is usually missed on custom inputs. Depends on the implementation and the use of native input, it's very unlikely that it will dispatch the events like native input does and the reason why we can't setup a change / blur event listener on the form element and trigger a form submit whenever the user interactive with a custom input. My goal with `useInputEvent()` hook is to progressively enhance custom inputs by emulating the event dispatched by native inputs.

Now, let's get back to your questions.

> 1. Why use useInputEvent if I could just use useState() to update the hidden input?

It depends on the setup. When you use `control.change()` from `useInputEvent()`, it does try to update the input value for you. If you own the hidden input, you could just set a default value to the hidden input and have Conform updating the hidden input value for you without a `useState`.

```tsx
function Exmaple() {
  const inputRef = useRef(null);
  const control = useInputEvent({
    ref: inputRef,
  });

  return (
    <>
      <input name="something" ref={inputRef} defaultValue="" hidden>
      <CustomInput onChange={control.change} /> 
    </>
  );
}
``` 

However, this could be risky if the hidden input is managed by the CustomInput and mess up its internal state. 

> 2. When and why to use each control method? E.g. control.change, control.focus and control.blur. Because I noticed in the official examples, not all control methods are used in all integrated components.

`control.focus()` is optional. What Conform concerns is only the two events mentioned above. Depends on your form configuration with `shouldValidate` and `shouldRevalidate`, you can also skip `control.blur()` if you never need to validate `onBlur`.

> 3. What's the differences of the onFocus handler in useInputEvent and the onFocus property in the referenced input? (Vice versa for onChange and onBlur)

If you mean hidden input by `referenced input`, it's the same. They are both listening to native events. The handler on the `useInputEvent` could be useful when you don't have access to the hidden input. (e.g. it is rendered by the custom input you have no control)

---

Thanks for all the questions. They are totally valid and I can see why the current API is confusing. It is expecting the developer to have some basic ideas on how the input is implemented or rendered and then figure out how to improve it.

I will try to come up with a different approach to simplify this process before v1. But I think the integration would also get simpler eventually as the react community continues to explore event delegation and hopefully takes native events as part of their contract. Also, we will be able to build more complex input with native element like [selectmenu](https://www.smashingmagazine.com/2023/06/advanced-form-control-styling-selectmenu-anchoring-api/) as the platform gets better and there is no need to do any integration one day.
